### PR TITLE
jmap_core: update all the Blob APIs to the latest draft-ietf-jmap-blob-04 draft

### PR DESCRIPTION
**Pull request by brong**
This changes Blob/get to Blob/lookup, and adds the new Blob/get and Blob/set logic from the IETF draft.

### CODE REVIEW

File: imap/jmap_core.c
Comment (rsto) [inline imap/jmap_core.c:461]: Both "offset" and "length" arguments should be checked for negative values.

File: imap/jmap_core.c
Comment (rsto) [inline imap/jmap_core.c:514]: I assume that's copied from old code, still I think that `-1` isn't a good choice here. Instead it should be set to HTTP_SERVER_ERROR or any more appropriate error. In this particular case wouldn't it even result in an `invalidArguments` JMAP error?

Remark: for new JMAP methods, I even tend to make them return `void` and take a `json_t **err` argument. That way I force the method handler to take care or error reporting.

File: imap/jmap_core.c
Comment (rsto) [inline imap/jmap_core.c:575]: See my comment above, we might want to `jmap_error` here in case of error.

**Review by rsto — CHANGES_REQUESTED** [review]
Looks good, with a few caveats. Argument validation could be improved. Mainly though, it seems to me that this patch only works for regular `G` blobs? Should the `lookup` method also use getblob_ctx to allow for handling smart blob ids? Probably I just don't see it...

File: imap/jmap_core.c
Comment (brong) [reply imap/jmap_core.c:514]: My plan here was that if there's an error, we add the blob to notFound, on the grounds that there was no blob data found within the specified range.  It would also be possible to just return `null` for the data bits if the range doesn't intersect but the blob does exist - I think that's reasonable.

File: imap/jmap_core.c
Comment (brong) [reply imap/jmap_core.c:575]: I think if we parsed OK, then unless there's a really bad error, we just want to return notFound, but... yeah, maybe we could bail and error if there's a real system error or something.

**Comment by brong** [comment]
Many updates pushed!  @rsto I've also pushed an updated copy of the RFC, and an updated copy of the MR for fastmail hm to match the new spec.

**Comment by brong** [comment]
@rsto two bugfixes as found by https://github.com/cyrusimap/cassandane/pull/173.  I'll keep revising both these PRs to actually match the spec before merging them, but build the the intermediate packages for Fastmail so we can convert the middleware first.

**Comment by brong** [comment]
Once again, I've churned this based on list feedback.  

@rsto, I'll ask for a final review once I've rebased everything together, but let's wait until next week for feedback at the interim.

**Comment by brong** [comment]
This is now in working group last call :)  Looks like it's unlikely to change

**Comment by brong** [comment]
Gonna merge this so we can rebase other work on top - if there are any more changes, we can do them later.

**Pull request merged**
